text/0000-execution-environments.md

```diff
@@ -0,0 +1,245 @@
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Separate Execution Environment Project Descriptor files
+
+When using `Dockerfile`, it's common to create a separate one for other execution environments like `Dockerfile.test`. This makes the file clean and easy to read, but comes at the cost of duplicating setup/code.
+
+With Project Descriptor, this would remove the need for the `exec-env` key at the cost of replicating buildpack groups. It does bring a big benefit by easily unlocking the `builder` key in Project Descriptor to be specific to an execution environment. This would apply to any fields (not tables) that exist at the `[io.buildpacks]` level like `include`/`exclude`. There are some other options if this is desired, where a new table could be created for hoisting fields into that would allow us to add the `exec-env` field.
+
+# Prior Art
+[prior-art]: #prior-art
+
+## [Heroku Testpack API](https://devcenter.heroku.com/articles/testpack-api)
+
+As part of the classic buildpack design, Heroku extended the API to include support for tests. It added a `bin/test-compile` phase, which is like the normal `bin/compile` phase but specific for setting things up for test. `bin/test` was called for executing tests. This design let a buildpack have a specific codepath for tests and also do intelligent logic for executing tests based on the source code. For example [in Ruby](https://github.com/heroku/heroku-buildpack-ruby/blob/main/bin/support/ruby_test#L62-L76), the buildpack author can run `bin/rspec` if it detects `rspec`, `bin/rails test` when using rails, or default to `rake test`.
+
+There were some flaws in this design. Though it's clean to separate production and test code paths, they end up sharing a lot of code. Many of the bash based Heroku buildpacks would just [call `bin/compile`](https://github.com/heroku/heroku-buildpack-nodejs/blob/main/bin/test-compile#L24) with different parameters/env vars.
+
+## [GOOGLE_DEVMODE](https://cloud.google.com/docs/buildpacks/service-specific-configs#google_devmode)
+
+These are specific to the Google Cloud Buildpacks for setting a development code path to work with skaffold.
+
+## [Develop API](https://github.com/buildpacks/spec/pull/71)
+
+The original Cloud Native Buildpacks spec included a Develop API, but it was never implemented.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- "env" is overloaded as a word since we also use it for environment variables. Is there a better word here?
+- Should there be builders that are specific to an execution environment? What about `include` or `exclude`?
+- Should the execution environments be an enum or flexible as a string?
+  - enums will help encourage standardization across buildpacks and platforms.
+  - strings can help account for use cases we haven't thought of yet.
+- Should buildpacks be allowed specify allowlist execution environments?
+- What changes are needed in the buildpack registry?
```

Comment:
Users will probably want a way to tell if a buildpack supports execution environments and if so, which ones. Maybe searching by that too. i.e. filter by buildpacks that support execution environments.

Reply:
@jkutner should we make changes to the registry index to account for this somehow? Add it into `buildpack.toml`?